prometheus_pgbouncer_exporter/cli.py

```diff
@@ -46,30 +74,24 @@ def _on_sigterm(signal, frame):
     signal.signal(signal.SIGTERM, _on_sigterm)
 
     # Read config file
-    config = Config()
-    try:
-        config.read(args.config)
-        logging.getLogger().info("Config file successfully read from {file}".format(file=args.config))
-    except Exception as error:
-        logging.getLogger().fatal("Unable to read config file from {file}".format(file=args.config), extra={"exception": str(error)})
-        sys.exit(1)
-
-    # Validate config
-    try:
-        config.validate()
-    except Exception as error:
-        logging.getLogger().fatal("The config file {file} is invalid: {error}".format(file=args.config, error=str(error)))
-        sys.exit(1)
+    config = read_config_file(args.config)
 
     # Register our custom collector
-    REGISTRY.register(PgbouncersMetricsCollector(config.getPgbouncers()))
+    pgbcollector = PgbouncersMetricsCollector(config.getPgbouncers())
+    REGISTRY.register(pgbcollector)
 
     # Start server
     start_http_server(config.getExporterPort(), config.getExporterHost())
     logging.getLogger().info("Exporter listening on {host}:{port}".format(host=config.getExporterHost(), port=config.getExporterPort()))
 
     while not shutdown:
-        time.sleep(1)
+        # check if HUPs were received in the last cycle and subsequently reload the pgbouncer hosts
+        if sighup_received_count:
+            logging.getLogger().info("Processing SIGHUP - Reloading pgbouncer hosts")
+            sighup_received_count = 0
+            config = read_config_file(args.config)
+            pgbcollector.update(config.getPgbouncers())
+        time.sleep(4)
```

Comment:
Why are you changing the sleep time? Is there any rationale behind it?

Reply:
While earlier the while loop would just sleep for a second upon check for shutdown, it is now also doing a  config reload if there were non-zero HUP counts during the sleep period. We have over 50 pgbouncer services running at a time that dynamically come and go. At times, especially during deployment rollout, the pgbouncer_exporter gets a flurry of HUP signals from a huge # of pgbouncer services as they get registered/deregistered in consul. A 1 sec sleep period ended up causing unnecessary thrashing due to excessive number of config reloads, (about 1 every second). As we already keep track of the counts of HUP that the services receives in during a while loop execution, we didn't need to be this aggressive in our reload. The logs also appear cleaner in that time period and easier to follow. A 4 sec sleep interval worked well for our purposes.

Reply:
Any `sleep(3)` added to the HUP handler can end in nested handling for subsequent HUP signals that are received during this sleep interval (HUP signal handling from within HUP signal handling). For cases when we can receive 4 to 5 HUPs per second, the whole service might end up stalling for  12-15 sec at a time. This can also end up in weird race condition as newer HUPs are processed first and then older ones (though that probably won't be an issue here as we are just updating a non-local variable). This  is the main reason we decided to just do a minimum work of adding HUP counts and quickly exit as part of HUP handling. 

We  could just go with `time.sleep(1)`. It doesn't affect the core functionality in any way. It just means that two successive config reloads can happen as quickly as 1 sec apart, instead of 4 seconds, if there is a high frequency of incoming HUPs, which on some rare occasions could add a slight strain on the service but is probably ok.

Reply:
Ok, seems fair enough. I don't have any other doubt about it. Thank you for explanation 😄